Config explainer (#125)
populationgenomics/team-docs: cgp-utils_config.md

# Configuration

> The [cpg-utils library](https://github.com/populationgenomics/cpg-utils) ([pypi](https://pypi.org/project/cpg-utils/)) contains a streamlined config management tool. This config management is used by the seqr-loader pipeline, but is useful in projects at any scale.

Reviewer: "This config management is used by the seqr-loader pipeline" would date pretty quickly. It is already used by other production workflows (e.g. large-cohort workflow, and under-the-construction-gatk-sv), as well as by some other analysis scripts we are running across the teams. Maybe rephrase 🤔

> The [cpg-utils library](https://github.com/populationgenomics/cpg-utils) ([pypi](https://pypi.org/project/cpg-utils/)) contains a streamlined config management tool. This config management is used by most production CPG workflows, but is useful in projects or scripts at any scale.

Reviewer 2: ```suggestion
The [`cpg-utils` library](https://github.com/populationgenomics/cpg-utils) ([pypi](https://pypi.org/project/cpg-utils/)) contains a streamlined config management tool. This config management is used by most production CPG workflows, but is useful in projects or scripts at any scale.
```

This configuration tool uses one or more `TOML` files, and creates a dictionary of key-value attributes which can be accessed at any point, without explicitly passing a configuration object.

> ## TOML

Reviewer: Nice! Love a quick intro to TOML syntax

[Tom's Obvious, Minimal Language](https://toml.io/en/) is a config file format designed to be easily human readable and writeable, with clear data structures. Sections are delineated using bracketed headings, and key-value pairs are defined using `=` syntax, e.g. :

```toml
global_key = "value"

[heading_1]
name = "Luke Skywalker"
age = 53

[heading_1.subheading]
occupation = ["Jedi", "Hermit", "Force Ghost"]
```

will be digested into the simple dictionary:

```python
{
    'global_key': 'value',
    'heading_1': {
        'name': 'Luke Skywalker',
        'age': 53,
        'subheading': {
            'occupation': ["Jedi", "Hermit", "Force Ghost"]
        }
    }
}
```

> ## Using cpg-utils.config

Reviewer 2: ```suggestion
## Using `cpg-utils.config`
```

To use the `cpg-utils.config` functions, import these two key methods into any script:

```python

> from cpg_utils.config import get_config, set_config_paths

Reviewer: You generally don't need to call `set_config_paths` as analysis-runner does that for you

```

Configuration is loaded lazily, so config files are loaded into memory only when we try to retrieve values from them. This reduces start-up overhead for scripts using this config approach, but can result in late failures if files with invalid content are specified.

> ### Setting configuration content

Reviewer: If we are committing to analysis-runner, we don't need to worry about `CPG_CONFIG_PATH` and `set_config_paths`, but instead just pass configs to analysis-runner with multiple `--config` options, and use `copy_common_env` if we want to pass a config down a Batch job.

Reviewer: It's still valuable though, in terms of describing how it works under the hood. So maybe move this section under a "development""in depth"/"under the hood" sub-topic?

Author: relevant to this comment and the one above, MSFT are interested in separating AIP from dependence on analysis-runner. This guide was going through the mechanics of config.py, with a section at the end addressing how analysis-runner will handle most of these operations for you

Reviewer 2: Maybe it would be helpful to split this into some sort of every-day usage, particularly within CPG, and "advanced" use cases like yours where you can't rely on the analysis-runner? I'd agree with Vlad that otherwise this might be a little confusing for our internal use cases, where you'd never need to call `set_config_paths` explicitly.

Configuration files can be set in two ways:

1. Set the `CPG_CONFIG_PATH` environment variable
2. Use `set_config_paths` to point to one or more config TOMLs

> `CPG_CONFIG_PATH` can be a comma-delimited String, and the argument to `set_config_paths` is a list of `strings`. The

Reviewer 2: ```suggestion
`CPG_CONFIG_PATH` can be a comma-separated list of config files, and the argument to `set_config_paths` is a list of `strings`. The
```

config functionality allows for multiple files to be specified, applied in the following way:

> 1. The cpg-utils repository contains a template config,

Reviewer 2: Also worth mentioning that the template contains descriptions of the parameters as comments. If anybody adds a generically useful config parameter, the template "documentation" should also be updated.

[toml template](https://github.com/populationgenomics/cpg-utils/blob/main/cpg_utils/config-template.toml); this is loaded first and becomes the base config.
2. For each additional config file in order, recursively update the base config with further content. New content is added, and content with the exact same key is updated/replaced.

> #### config loading example

Reviewer 2: ```suggestion
#### Config loading example
```

### Getting configuration content

Once configuration paths are set, values can be retrieved at any point using `get_config`. The first call to `get_config` sets the global config dictionary and returns the content, subsequent calls will just return the config dictionary.

```python
assert get_config()['file'] == 'second.toml'
```

## Analysis-Runner

Analysis-runner incorporates a simple interface for config setting. When setting off a job, the flag `--config` can be used, pointing to a config file (local, or within GCP and accessible with current logged-in credentials).
The `--config` flag can be used multiple times, which will cause the argument files to be aggregated in the order they are defined. When `--config` is set in this way, the job-runner performs the following actions:

1. Locally (where `analysis-runner` is invoked), a [merged configuration file is generated](https://github.com/populationgenomics/analysis-runner/blob/main/analysis_runner/cli_analysisrunner.py#L199-L201), creating a single dictionary
2. This dictionary is sent with the job definition to the execution server

> 3. The data is saved in TOML format to a dataset-specific temporary location in GCP

Reviewer 2: Might be worth mentioning that this is helpful for reproducible runs (as e.g. local configs that are not committed to `git` are also supported, as you already mention above). The path to the merged config is logged in Hail Batch, at least for `analysis-runner` invocations.

Reviewer 2: We do store the merged configs permanently in `gs://cpg-config`: https://github.com/populationgenomics/analysis-runner/blob/ba826f574f1a1049ce01bc7236073bdbd3b3a0ec/server/util.py#L168-L173

Author: Ah sorry 🤦  I thought I remembered the file being written to -tmp! 🤦🤦
